### ISSUE TITLE
How to use the rope scaling with x_transformers ?

### ISSUE THREAD

**Issue by cutoken** [lucidrains/x-transformers#159]
Hi,
Thank you for this wonderful library. I'm wondering if there is a way to use the recent rope paper's scaling workaround with x_transformers. I have seen your recent change to the rotary position encoding repo but wasn't able to identify where to modify similarly in x_transfomers repo.

**Comment by lucidrains** [comment]
@cutoken oh hey

do you mean the interpolation of the positions? do you have a pretrained model with RoPE you are finetuning?

**Comment by lucidrains** [comment]
you are referring to https://github.com/lucidrains/rotary-embedding-torch/commit/e7ce8e0bee8b09cbba835cd67edc666ada525c40 ?

**Comment by cutoken** [comment]
It is actually quite weird. I'm able to recreate the issue of increasing loss on context increase but the interpolation solution doesn't really work for me. Have you had success recreating the paper on any smaller model ?

**Comment by lucidrains** [comment]
@cutoken do you mean you tried it just now and it didn't work? did you follow their recipe of fine tuning on 1k longer context samples, with an `rotary_interpolation_factor = new_max_seq_len / old_max_seq_len` ?

**Comment by lucidrains** [comment]
@cutoken if i hear it doesn't work from a few people, i may just remove it in favor of xpos

**Comment by cutoken** [comment]
Will confirm once again in a fresh experiment @lucidrains

**Comment by lucidrains** [comment]
share it with w&b!

**Comment by cutoken** [comment]
Can confirm it is actually worse than  directly fine tuning without interpolation. I'm using a really small model but I don't see why that should matter as the interpolation factor is being followed as per the paper guidelines. I can send you the weights and biases if needed of the smaller pre-intrapolated one so that you can also test if needed (please provide your mail id in that case)

**Comment by lucidrains** [comment]
@cutoken that would be great! could you share the training script too? i think i'll go ahead and remove it from this repository until i hear more feedback (or see a paper that corroborates the technique)

**Comment by lucidrains** [comment]
@cutoken do you want to double check your experiments. i see someone legit corroborating the results from Meta https://kaiokendev.github.io/context

**Comment by lucidrains** [comment]
@cutoken they also released a new long context model, LongChat, at 16k

**Comment by cutoken** [comment]
I hope it works as well  :)

Below link contains the training script I have used and also the checkpoint with rope enabled (both checkpoints are the same. Just made a backup in case you end up overwriting it). You can use it as the starting point to train with and without the newly added parameter to see the difference. You will need TinyStories data set from huggingface. I have added the sentencepiece vocab file already so you would only need sentencepiece available - no need to tokenize the dateset again. Let me know if you face any issues running it.

[warning - large file due to the model size](https://www.transfernow.net/dl/20230630VETStt8r)

**Comment by lucidrains** [comment]
@cutoken thanks! i'll allot some time this Sunday to do some training

**Comment by lucidrains** [comment]
@cutoken if you read [their blog post](https://lmsys.org/blog/2023-06-29-longchat/), it seems like the technique should work! this group def knows what they are doing. threw a bit of shade at MosaicML too haha

**Comment by kaiokendev** [comment]
Hello, it is really cool you added this feature! One thing I would mention is that in my case, the intuition is that the large model may overfit to the position embedding, such that it is easier to train on the interpolated position than using OOD positions. The counter point is that small models may not be overfit in the same way - I see Meta only trained on 7B parameters and up, so it's possible the effect decreases for smaller models. There was also no ablation performed for non-LLaMA RoPE models, so it is unknown how much it depends on other factors as well. Just a thought

**Comment by cutoken** [comment]
@kaiokendev , that sounds like a good explanation on why I'm not seeing  same results with a smaller model (50M params).

**Comment by lucidrains** [comment]
@kaiokendev oh interesting; you can probably run a few experiments to back up your idea, and share it on twitter

would be an important caveat that should be noted in their paper!

**Comment by lucidrains** [comment]
@kaiokendev there's actually a number of papers popping up here and there that tries to reduce overfitting of the positional embeddings

the two i've seen are (1) randomly offset positions by some constant and (2) within a range of 0 to a length L where L > maximum number of tokens, use a random subset of positions from that range, ascending

**Comment by lucidrains** [comment]
@kaiokendev  have you seen https://arxiv.org/abs/2307.03170 ?

**Comment by kaiokendev** [comment]
> @kaiokendev have you seen https://arxiv.org/abs/2307.03170 ?

I did skim it, there are a lot of external memory approaches I saw, but I do not really play with the cases involving approximated attention like that one
